sys/kerncore/src/lib.rs

```diff
@@ -0,0 +1,389 @@
+
+    /// The address of the first byte _not_ included in this slice, past the
+    /// end.
+    ///
+    /// Note that this prevents a slice from touching the end of the address
+    /// space. This is also prevented, in practice, by the definition of several
+    /// Rust core types, so we accept it.
+    ///
+    /// The return value must be greater than or equal to the result of
+    /// `base_addr`.
+    fn end_addr(&self) -> usize;
+}
+
+impl<T: UserSlice> UserSlice for &T {
+    #[inline(always)]
+    fn is_empty(&self) -> bool {
+        (**self).is_empty()
+    }
+
+    #[inline(always)]
+    fn base_addr(&self) -> usize {
+        (**self).base_addr()
+    }
+
+    #[inline(always)]
+    fn end_addr(&self) -> usize {
+        (**self).end_addr()
+    }
+}
+
+/// Describes types that indicate region permissions.
+///
+/// This type is _almost_ identical to `UserSlice` but has slightly different
+/// operations defined on it. Those operations _do not_ include region
+/// attributes, which might be surprising, but we handle those generically using
+/// a predicate when required.
+///
+/// # Invariants
+///
+/// `r.contains(x)` implies `r.base_addr() <= x < r.end_addr()` and vice versa.
+///
+/// By extension, `r.base_addr() <= r.end_addr()` must hold.
+///
+/// An empty region is weird but not impossible.
+pub trait MemoryRegion {
+    fn contains(&self, addr: usize) -> bool;
+    fn base_addr(&self) -> usize;
+    fn end_addr(&self) -> usize;
+}
+
+impl<T: MemoryRegion> MemoryRegion for &T {
+    #[inline(always)]
+    fn contains(&self, addr: usize) -> bool {
+        (**self).contains(addr)
+    }
+
+    #[inline(always)]
+    fn base_addr(&self) -> usize {
+        (**self).base_addr()
+    }
+
+    #[inline(always)]
+    fn end_addr(&self) -> usize {
+        (**self).end_addr()
+    }
+}
+
+/// Generic version of the kernel slice access checking code.
+///
+/// The purpose of this routine is to determine whether a task can access some
+/// memory. The memory is described by `slice` and consists of a single
+/// contiguous region. The task's memory access permissions are described by
+/// `table`, which is an array of region descriptors.
+///
+/// The exact implementation of both the slice type `S` and the region type `R`
+/// are left unspecified here, to avoid needing to rely on kernel-internal
+/// types. The requirements for each type are specified by [`UserSlice`] and
+/// [`MemoryRegion`], respectively.
+///
+/// Generally speaking, not all task region descriptors confer the same rights
+/// --- some are read-only, some may represent an address space hole that cannot
+/// be accessed, etc. To reflect this, this routine takes a `region_ok`
+/// predicate over the `R` type. Provide a `region_ok` function to do any access
+/// permission checking that you require.
+///
+/// # Preconditions
+///
+/// `table` must be sorted by region base address, and the regions in the table
+/// must not overlap.
```

Comment:
We don't build the _kernel_ with debug assertions, no. I thought about this, and it would find bugs in the test fixture in this crate, which is something? But it wouldn't detect bugs in the kernel.